### PR TITLE
Impl ChunkData (0x21) + Chunk and ChunkColumn types.

### PR DESCRIPTION
**Pull request by toqueteos**
Ported from https://github.com/toqueteos/hematite_server_ugly.

I would like some opinions especially on decoding. How can decode know what dimension the player is located on? Static server info available to anyone everywhere?

**EDIT:** Forgot reference links!
- http://wiki.vg/Protocol#Chunk_Data
- http://wiki.vg/SMP_Map_Format#Data

/cc @eddyb @fenhl

### CODE REVIEW

File: src/packet.rs [hunk 1]
```diff
@@ -431,7 +431,96 @@ packets! {
...
+                                *x = try!(<u16 as Protocol>::proto_decode(src));
+                            }
+                        }
+                        for chunk in column.chunks.iter_mut() {
+                            // We use this instead of read_exact because it's an array, Vec is useless here.
+                            for x in chunk.block_light.iter_mut() {
+                                *x = try!(<u8 as Protocol>::proto_decode(src));
+                            }
+                        }
+                        // FIXME(toqueteos): Tricky one, only sent if player is in Overworld.
```
Comment (fenhl) [inline src/packet.rs]: We might have to save the raw chunk column data, and give it a decode method that takes the dimension.

Comment (toqueteos) [inline src/packet.rs]: Post-processing after reading the packet?

Comment (fenhl) [inline src/packet.rs]: Essentially, yes.

Comment (toqueteos) [inline src/packet.rs]: Almost done!

**Comment by toqueteos** [comment]
I was doubting if adding `Dimension` or not... Seems better with it in.

File: src/types/chunk.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        for chunk in self.chunks.iter() {
+            for x in chunk.blocks.iter() {
+                try!(<u16 as Protocol>::proto_encode(x, &mut dst));
+            }
+        }
+        for chunk in self.chunks.iter() {
+            try!(dst.write_all(&chunk.block_light));
+        }
+        for chunk in self.chunks.iter() {
+            match chunk.sky_light {
```
Comment (toqueteos) [inline src/types/chunk.rs:39]: Should we check for Dimension here too (encode)? What if player is on Overworld but `chunk.sky_light` is `None`?

Comment (toqueteos) [inline src/types/chunk.rs:39]: ~~Sidenote: `encode`/`decode` comments are terrible. Will fix tomorrow.~~ FIXED

**Comment by fenhl** [comment]
Since this adds `Dimension`, it should also change the types of the `dimension` fields in the `JoinGame` and `Respawn` packets.

**Comment by toqueteos** [comment]
Done. I made the `types::consts` module pub so we can use `use types::consts::*`.

**Comment by fenhl** [comment]
`Dimension` is missing a `Protocol` impl.

**Comment by toqueteos** [comment]
D'oh! Manual or macro? Something like `impl_protocol!` would do...

**Comment by toqueteos** [comment]
C-like enums are such a joy to deal with!

File: src/types/consts.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+use packet::Protocol;
+
+macro_rules! enum_protocol_impl {
+    ($name:ty, $len:expr, $repr:ty, $dec_repr:ident) => {
+        impl Protocol for $name {
+            type Clean = $name;
+
+            #[allow(unused_variables)]
+            fn proto_len(value: &$name) -> usize { $len }
```
Comment (fenhl) [inline src/types/consts.rs]: Why not just `<$repr as Protocol>::proto_len(&(*value as repr))`?

Comment (toqueteos) [inline src/types/consts.rs]: That's nice, one less macro param!

**Comment by toqueteos** [comment]
Ping @fenhl

**Comment by fenhl** [comment]
*yawn* hmm?

**Comment by toqueteos** [comment]
Is there something else to change?

**Comment by fenhl** [comment]
I don't think so.

**Comment by toqueteos** [comment]
Merging, we can now move forward with #8.

**Pull request merged**